toplevel/genprintval.ml

```diff
@@ -388,7 +388,8 @@ module Make(O : OBJ)(EVP : EVALPATH with type valu = O.t) = struct
                 | {type_kind = Type_variant (constr_list,rep)} ->
                     let unbx = (rep = Variant_unboxed) in
                     let tag =
-                      if unbx then Cstr_unboxed
+                      if unbx then (* XXX unboxing Cstr_unboxed [] *)
+                        assert false
                       else if O.is_block obj
                       then Cstr_block(O.tag obj)
                       else Cstr_constant(O.obj obj) in
```

Comment:
This genprintval.ml adaptation is tricky and will require more work. What the code is doing is to use type information to print a "raw value" in the toplevel. For boxed constructors, we use `Obj.tag` to get the tag at runtime, and then we query the constructor-descriptions to get the expected arity and type of the argument(s).

What should we do with unboxed constructor? What we need in general is a mapping from the tag (or immediate value) of the runtime value `obj` to one specific constructor, and to do this we need to inspect head shapes: if the value comes from an unboxed constructor, we can find it by comparing the runtime tag with its head shape.

Note: this does not work well for unboxed constructor whose argument is an empty type, if we allow several such constructors. But this is not a problem because such unboxed constructors are un-inhabited, the runtime value we have will never come from one of those (non-disjoint) unboxed constructors anyway.

Reply:
Discussing this again: things would work correctly if `Datarepr.find_constr_by_tag` was aware that the "raw tag" or "runtime tag" passed (maybe we should use a type different from `constructor_tag`; `runtime_constructor_tag`, without an `unboxed` case?) may not be one of the boxed tags, but it may also belong to the head shape of (at most one of) the unboxed constructors.

(We may need to force the `head_shape` here, instead of within `Datarepr`.)

Reply:
@nchataing remarks that this should be computed by matching not on the variant type declaration, but on the type description.